internal/execution/handlers.go

```diff
@@ -0,0 +1,177 @@
+
+	"github.com/zitadel/zitadel/internal/domain"
+	"github.com/zitadel/zitadel/internal/eventstore"
+	"github.com/zitadel/zitadel/internal/eventstore/handler/v2"
+	"github.com/zitadel/zitadel/internal/query"
+	"github.com/zitadel/zitadel/internal/queue"
+	exec_repo "github.com/zitadel/zitadel/internal/repository/execution"
+)
+
+const (
+	HandlerTable = "projections.execution_handler"
+)
+
+type eventHandlerQueries interface {
+	TargetsByExecutionID(ctx context.Context, ids []string) (execution []*query.ExecutionTarget, err error)
+}
+
+type Queue interface {
+	Insert(ctx context.Context, args river.JobArgs, opts ...queue.InsertOpt) error
+}
+
+type eventHandler struct {
+	eventTypes                 []string
+	aggregateTypeFromEventType func(typ eventstore.EventType) eventstore.AggregateType
+	query                      eventHandlerQueries
+	queue                      Queue
+}
+
+func NewEventHandler(
+	ctx context.Context,
+	config handler.Config,
+	eventTypes []string,
+	aggregateTypeFromEventType func(typ eventstore.EventType) eventstore.AggregateType,
+	query eventHandlerQueries,
+	queue Queue,
+) *handler.Handler {
+	return handler.NewHandler(ctx, &config, &eventHandler{
+		eventTypes:                 eventTypes,
+		aggregateTypeFromEventType: aggregateTypeFromEventType,
+		query:                      query,
+		queue:                      queue,
+	})
+}
+
+func (u *eventHandler) Name() string {
+	return HandlerTable
+}
+
+func (u *eventHandler) Reducers() []handler.AggregateReducer {
+	aggList := make(map[eventstore.AggregateType][]eventstore.EventType)
+	for _, eventType := range u.eventTypes {
+		aggType := u.aggregateTypeFromEventType(eventstore.EventType(eventType))
+		aggEventTypes, ok := aggList[aggType]
+		if !ok {
+			aggList[aggType] = []eventstore.EventType{eventstore.EventType(eventType)}
+		} else {
+			found := false
+			for _, aggEventType := range aggEventTypes {
+				if aggEventType == eventstore.EventType(eventType) {
+					found = true
+				}
+			}
+			if !found {
+				aggList[aggType] = append(aggList[aggType], eventstore.EventType(eventType))
+			}
+		}
+	}
+
+	aggReducers := make([]handler.AggregateReducer, len(aggList))
+	i := 0
+	for aggType, aggEventTypes := range aggList {
+		eventReducers := make([]handler.EventReducer, len(aggEventTypes))
+		for j, eventType := range aggEventTypes {
+			eventReducers[j] = handler.EventReducer{
+				Event:  eventType,
+				Reduce: u.reduce,
+			}
+		}
+		aggReducers[i] = handler.AggregateReducer{
+			Aggregate:     aggType,
+			EventReducers: eventReducers,
+		}
+		i++
+	}
+	return aggReducers
+}
+
+func groupsFromEventType(s string) []string {
+	parts := strings.Split(s, ".")
+	groups := make([]string, len(parts))
+	groupBase := ""
+	for i, part := range parts {
+		if groupBase == "" {
+			groupBase = part
+		} else {
+			groupBase = groupBase + "." + part
+		}
+
+		if groupBase == s {
+			groups[i] = groupBase
+		} else {
+			groups[i] = groupBase + ".*"
+		}
+	}
```

Comment:
This code feels very complex for the task it is solving, isn't there a function in the strings package which makes the code more readable?

Reply:
Not that I know of, the thing is here with the best match, for the query there are as many possible values as there are elements in the group definition. 
For example for `user.human.mfa.init.skipped`:
- all
- user
- user.human
- user.human.mfa
- user.human.mfa.init
- user.human.mfa.init.skipped

Reply:
There is no direct function for it, but I reworked it with strings.join